HidPkg/Crates/HiiKeyboardLayout/src/lib.rs

```diff
@@ -0,0 +1,782 @@
+}
+
+/// HII Keyboard Layout
+/// Refer to UEFI spec version 2.10 section 34.8.10 which defines the keyboard layout structure.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct HiiKeyboardLayout {
+  /// The unique ID associated with this keyboard layout.
+  pub guid: efi::Guid,
+  /// A list of key descriptors
+  pub keys: Vec<HiiKey>,
+  /// A list of descriptions for this keyboard layout.
+  pub descriptions: Vec<HiiKeyboardDescription>,
+}
+
+/// HII Key descriptor
+/// Refer to UEFI spec version 2.10 section 34.10.10 which defines the key descriptor structure.
+#[derive(Debug, Pread, Pwrite, PartialEq, Eq, Clone, Copy)]
+#[repr(C)]
+pub struct HiiKeyDescriptor {
+  /// Describes the physical key on the keyboard.
+  pub key: EfiKey,
+  /// Unicode character for the key (note: UEFI only supports UCS-2 encoding).
+  pub unicode: u16,
+  /// Unicode character for the key with the shift key being held down.
+  pub shifted_unicode: u16,
+  /// Unicode character for the key with the Alt-GR being held down.
+  pub alt_gr_unicode: u16,
+  /// Unicode character for the key with the Alt-GR and shift keys being held down.
+  pub shifted_alt_gr_unicode: u16,
+  /// Modifier keys are defined to allow for special functionality that is not necessarily accomplished by a printable
+  /// character. Many of these modifier keys are flags to toggle certain state bits on and off inside of a keyboard
+  /// driver. See [`r_efi::protocols::hii_database`] for modifier definitions.
+  pub modifier: u16,
+  /// Indicates what modifiers affect this key. See [`r_efi::protocols::hii_database`] for "affected by" definitions.
+  pub affected_attribute: u16,
+}
+
+/// Non-Spacing HII Key Descriptor variant. Used for "non-spacing" keys.
+/// Refer to discussion in UEFI spec version 2.10 section 33.2.4.3 for information on "non-spacing" keys and how they
+/// are used.
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct HiiNsKeyDescriptor {
+  /// The descriptor for the "non-spacing key" itself.
+  pub descriptor: HiiKeyDescriptor,
+  /// The list of descriptors that are active if the "non-spacing" key has been pressed.
+  pub dependent_keys: Vec<HiiKeyDescriptor>,
+}
+
+/// HII Key descriptor enumeration.
+/// HII spec allows for two types of key descriptors - normal and "non-spacing".
+/// Refer to UEFI spec version 2.10 section 33.2.4.3
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum HiiKey {
+  Key(HiiKeyDescriptor),
+  NsKey(HiiNsKeyDescriptor),
+}
+
+/// Enumeration of physical keys.
+/// Refer to UEFI spec version 2.10 section 34.8.10 and section 33.2.4.1.
+#[rustfmt::skip]
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+//Note: UEFI specifies this as an C enum. That means the size is a bit ambiguous; but most compilers
+//will make it 32-bit, so that's what this implementation assumes.
+#[repr(u32)]
+pub enum EfiKey {
+  EfiKeyLCtrl, EfiKeyA0, EfiKeyLAlt, EfiKeySpaceBar, EfiKeyA2, EfiKeyA3, EfiKeyA4, EfiKeyRCtrl, EfiKeyLeftArrow,

```

Comment:
Do you think the enum members each need to be prefixed with `EfiKey` since they're in the `EfiKey` enum?

It makes usage like this `EfiKey::EfiKeyLCtrl` a bit repetitive.

Reply:
Implemented suggestion.